ichnaea/data/schema.py

```diff
@@ -0,0 +1,376 @@
+    from either a datetime object or a string.
+    """
+
+    def deserialize(self, schema, cstruct):
+        if type(cstruct) == datetime.datetime:
+            cstruct = cstruct.strftime('%Y-%m-%d')
+        return super(DateTimeToString, self).deserialize(schema, cstruct)
+
+
+# Custom Nodes
+
+class DefaultNode(SchemaNode):
+    """
+    A DefaultNode will use its ``missing`` value
+    if it fails to validate during deserialization.
+    """
+
+    def deserialize(self, cstruct):
+        try:
+            return super(DefaultNode, self).deserialize(cstruct)
+        except Invalid:
+            return self.missing
+
+
+class WifiKeyNode(SchemaNode):
+    """
+    A node containing a valid wifi key.
+    ex: 01005e901000
+    """
+
+    def preparer(self, cstruct):
+        return normalized_wifi_key(cstruct)
+
+    def validator(self, node, cstruct):
+        if not valid_wifi_pattern(cstruct):
+            raise Invalid(node, 'Invalid wifi key')
+
+
+class ReportNode(SchemaNode):
+    """
+    A node containing a valid report_id.
+    ex: 489cc8dc9d3d11e4a87d02442b52e5a0
+    """
+
+    def preparer(self, cstruct):
+        return cstruct or uuid.uuid1().hex
+
+
+class TimeNode(SchemaNode):
+    """
+    A node containing a valid date.
+    ex: 2015-01-01
+    """
+
+    def preparer(self, cstruct):
+        if cstruct:
+            return normalized_time(cstruct)
+        else:
+            return datetime.date.today().strftime('%Y-%m-%d')
+
+
+# Schemas
+
+class CopyingSchema(MappingSchema):
+    """
+    A Schema which makes a copy of the passed in dict to validate.
+    """
+
+    def deserialize(self, data):
+        return super(CopyingSchema, self).deserialize(copy.copy(data))
+
+
+class FieldSchema(MappingSchema):
+    """
+    A schema which provides an interface to its fields through the
+    .fields[field_name] interface.
+    """
+
+    @property
+    def fields(self):
+        return dict([(field.name, field) for field in self.children])
+
+    def is_missing(self, data, field_name):
+        missing_value = self.fields[field_name].missing
+        data_value = data.get(field_name, missing_value)
+        return data_value == missing_value
+
+
+class ValidMeasureSchema(FieldSchema, CopyingSchema):
+    """
+    A Schema which validates the fields present in a measurement,
+    regardless of whether it is a Cell or Wifi measurement.
+    """
+    lat = SchemaNode(Float(), missing=0.0, validator=Range(MIN_LAT, MAX_LAT))
+    lon = SchemaNode(Float(), missing=0.0, validator=Range(-180, 180))
+    accuracy = DefaultNode(
+        Float(), missing=0, validator=Range(0, MAX_ACCURACY))
+    altitude = DefaultNode(
+        Float(), missing=0, validator=Range(MIN_ALTITUDE, MAX_ALTITUDE))
+    altitude_accuracy = DefaultNode(
+        Float(), missing=0, validator=Range(0, MAX_ALTITUDE_ACCURACY))
+    heading = DefaultNode(Float(), missing=-1, validator=Range(0, MAX_HEADING))
```

Comment:
accuracy, altitude and altitude_accuracy should all be integers. The db fields are ints and the missing and ranges are all specified as ints.